### PR TITLE
approval-voting: simplify v2 assignments

### PR DESCRIPTION
**Pull request by sandreim**
This is a child PR of https://github.com/paritytech/polkadot/pull/6782.

Implementing some TODOs and removing redundant information as suggested in review discussion: https://github.com/paritytech/polkadot/pull/6782#discussion_r1117750373

Changes:
- Separate transcript for v1/v2 assignments
- remove `sample` and `core_indices` from v2 certificate.
- update `check_assignment_cert` to return the assigned cores as outputed by VRF.
- impl TODO: bitfield instead of `Vec<CandidateIndex>`
- impl TODO: bitfield instead of `Vec<CoreIndex>`

### CODE REVIEW

File: node/core/approval-voting/src/approval_db/v1/mod.rs [hunk 1]
```diff
@@ -166,6 +166,8 @@ pub struct OurAssignment {
 	pub validator_index: ValidatorIndex,
 	// Whether the assignment has been triggered already.
 	pub triggered: bool,
+	// Claimed core indices.
```
Comment (sandreim) [inline node/core/approval-voting/src/approval_db/v1/mod.rs:169]: this seems like a dumb comment.

File: roadmap/implementers-guide/src/node/approval/approval-voting.md [hunk 2]
```diff
@@ -54,6 +54,8 @@ struct OurAssignment {
   tranche: DelayTranche,
   validator_index: ValidatorIndex,
   triggered: bool,
+  // Claimed core indices.
```
Comment (sandreim) [inline roadmap/implementers-guide/src/node/approval/approval-voting.md:57]: this needs fixing.

File: node/core/approval-voting/src/approval_db/v1/mod.rs (hunk 1 above)
Comment (rphmeier) [reply node/core/approval-voting/src/approval_db/v1/mod.rs:169]: True :) 

Doc comments on pall ublic fields would be good

File: node/core/approval-voting/src/criteria.rs [hunk 3]
```diff
@@ -87,17 +90,41 @@ impl From<OurAssignment> for crate::approval_db::v1::OurAssignment {
...
+
+	transcript
+}
+
 fn relay_vrf_modulo_transcript(relay_vrf_story: RelayVRFStory, sample: u32) -> Transcript {
-	// combine the relay VRF story with a sample number.
-	let mut t = Transcript::new(approval_types::RELAY_VRF_MODULO_CONTEXT);
-	t.append_message(b"RC-VRF", &relay_vrf_story.0);
-	sample.using_encoded(|s| t.append_message(b"sample", s));
+	relay_vrf_modulo_transcript_inner(
```
Comment (rphmeier) [inline node/core/approval-voting/src/criteria.rs:114]: for consistency, rename this function to "_v1"

File: node/core/approval-voting/src/criteria.rs [hunk 4]
```diff
@@ -44,6 +44,8 @@ pub struct OurAssignment {
 	validator_index: ValidatorIndex,
 	// Whether the assignment has been triggered already.
 	triggered: bool,
+	// Claimed core indices.
+	pub claimed_core_indices: Vec<CoreIndex>,
```
Comment (rphmeier) [inline node/core/approval-voting/src/criteria.rs:48]: better practice to add a getter; mixing public and private fields always looks weird to me.

Can this legally be empty?

**Comment by burdges** [comment]
We've this security implication:  If we permit validators to claim less than all their assignments then the list of what they actually claim should really be signed by the extra transcript going into the VRF, or else signed by an external signature scheme, but even having an external signature scheme sounds wasteful.

At a higher level..

Do we need validators to be able to claim less than all their assignments?  It's unclear.  We do not have any mental model by which a validator could do this right now, and do not understand how doing this impacts security, but nothing prevents us inventing this in future, but maybe we should push validators onto common hardware instead.  Also, validators could simply skip some whole blocks.  We've never thought about the security implications of skipping whole blocks either of course.

tl;dr Do we leave a bitfield for functionality which we've no idea how to safely use anyways, but which morally sounds like load balancing somehow?

File: node/core/approval-voting/src/criteria.rs (hunk 4 above)
Comment (sandreim) [reply node/core/approval-voting/src/criteria.rs:48]: That's a good point. No it cannot be empty.

**Comment by sandreim** [comment]
> We've this security implication: If we permit validators to claim less than all their assignments then the list of what they actually claim should really be signed by the extra transcript going into the VRF, or else signed by an external signature scheme, but even having an external signature scheme sounds wasteful.
> 
> At a higher level..
> 
> Do we need validators to be able to claim less than all their assignments? It's unclear. We do not have any mental model by which a validator could do this right now, and do not understand how doing this impacts security, but nothing prevents us inventing this in future, but maybe we should push validators onto common hardware instead. Also, validators could simply skip some whole blocks. We've never thought about the security implications of skipping whole blocks either of course.
> 

This is a good point on which I spent some time thinking. Are we ok with the simplifying assumption that for now validators actually can claim a subset of the VRF output cores ? AFAIU this is not worse than current situation when they can not send out any assignments and skip whole blocks. We should think of ways to fix this when we implement incentives for approval checker work.

> tl;dr Do we leave a bitfield for functionality which we've no idea how to safely use anyways, but which morally sounds like load balancing somehow?

Yeah, load balancing is something that we will likely need, however I am inclined to not introduce such a bitfield yet as maybe not showing up because of load is also an option that would basically achieve the same thing.  I am a little bit worried TBH, since there is a chance that due to load some candidates might not get `needed_approvals` votes stalling finality.

**Comment by sandreim** [comment]
> We have this bitfield/list now as `claimed_core_indices: Vec<CoreIndex>` since @rphmeier did it this way the first time (and you wrote it into #6782 too), but do we want to permit them to take fewer assignments?
Until we implement approval checker incentivisation, it is no worse than what is currently possible. 

> 
> I have no intuition if skipping some assignments is worse than skipping whole blocks, or visa versa. It's likely easier to analyze whole blocks somehow, but not necessarily better for security. It's possible code elegance trumps other concerns now anyways. lol
> 

IMO, the worse case scenario would be that at least one candidate would remain unapproved and finality stalls, skipping a whole block has the same effect. 

> It's possible code elegance trumps other concerns now anyways. lol
> 

Might make sense for anything that is not production code :P

> If we permit taking only some assignments, then instead of the bitfield/list they have now we should let them give a number `accepted_assignments`, and then they take the assignments numbered `0..accepted_assignments`, which gets determined by the first `4*accepted_assignments` bytes of the VRF output. An adversary having less control maybe helpful somewhere, or this maybe simplifies our analysis, but adversary won't typically care about being no-show on an irrelevant core.
> 

I think this will be necessary probably for incentivisation, but I would rather not extend the initial scope of this change set.

**Comment by burdges** [comment]
I'll try to do a proper write up of incentivization but just fyi some notes exist at https://github.com/w3f/research-security-issues/issues/30#issuecomment-768593265 so message me on element if you do not have access.

**Comment by sandreim** [comment]
Seems that I need to update docs.

File: roadmap/implementers-guide/src/node/approval/approval-voting.md [hunk 5]
```diff
@@ -54,6 +54,8 @@ struct OurAssignment {
   tranche: DelayTranche,
   validator_index: ValidatorIndex,
   triggered: bool,
+	/// A subset of the core indices obtained from the VRF output.
+	pub claimed_core_indices: Vec<CoreIndex>,
```
Comment (ordian) [inline roadmap/implementers-guide/src/node/approval/approval-voting.md:58]: nit: md files use spaces instead of tabs

File: node/service/src/parachains_db/upgrade.rs [hunk 6]
```diff
@@ -278,6 +296,42 @@ fn paritydb_migrate_from_version_1_to_2(path: &Path) -> Result<(), Error> {
...
+		.to_str()
+		.ok_or_else(|| super::other_io_error("Invalid database path".into()))?;
+	let db_cfg = DatabaseConfig::with_columns(super::columns::v2::NUM_COLUMNS);
+	let db = Database::open(&db_cfg, db_path)?;
+	let entries = db.iter(super::columns::v2::COL_APPROVAL_DATA);
+
+	// TODO: find a more efficient way to wipe the column
+	let ops = entries
+		.filter_map(|result| result.ok())
+		.map(|(key, _)| DBOp::Delete { col: super::columns::v2::COL_APPROVAL_DATA, key })
```
Comment (ordian) [inline node/service/src/parachains_db/upgrade.rs:328]: maybe `DeletePrefix(col, &[])`?

File: node/primitives/src/approval.rs [hunk 7]
```diff
@@ -84,25 +196,20 @@ pub enum AssignmentCertKindV2 {
...
 	///
-	/// The context used to produce bytes is [`RELAY_VRF_MODULO_CONTEXT`]
+	/// The context used to produce bytes is [`v2::RELAY_VRF_MODULO_CONTEXT`]
 	RelayVRFModulo {
 		/// The sample number used in this cert.
 		sample: u32,
 	},
 	/// Multiple assignment stories based on the VRF that authorized the relay-chain block where the
-	/// candidate was included combined with a sample number.
+	/// candidate was included.
```
Comment (ordian) [inline node/primitives/src/approval.rs:205]: candidate**s** **were**?

File: node/core/approval-voting/src/lib.rs [hunk 8]
```diff
@@ -1049,26 +1064,23 @@ async fn handle_actions<Context>(
...
-					.iter()
-					.position(|(core_index, _)| core_index == cert_core_index)
-				{
-					candidate_indices.push(candidate_index as _)
-				}
-			}
-		},
-		_ => candidate_indices.push(candidate_index as _),
+
+	// Map from core index to candidate index.
```
Comment (ordian) [inline node/core/approval-voting/src/lib.rs:1072]: do we check the len of `candidate_indices` is the same as `core_indices.iter_ones().len()` or simply filter out garbage?

File: node/core/approval-voting/src/criteria.rs [hunk 9]
```diff
@@ -641,72 +674,81 @@ pub(crate) fn check_assignment_cert(
...
+			// Get unique core assignments from the VRF wrt `config.n_cores`.
+			let vrf_unique_cores = relay_vrf_modulo_cores(
+				&vrf_in_out,
+				config.relay_vrf_modulo_samples,
+				config.n_cores,
+			);
 
-			// Ensure that the `vrf_in_out` actually includes all of the claimed cores.
-			if claimed_core_indices
+			// Filter out cores in which the validator is in the backing group.
```
Comment (ordian) [inline node/core/approval-voting/src/criteria.rs:693]: why do we need this check now and didn't have this before? honest backing validators should not be participating in approvals

**Review by ordian — COMMENTED** [review]
some questions

File: node/core/approval-voting/src/criteria.rs [hunk 10]
```diff
@@ -231,14 +255,15 @@ pub(crate) trait AssignmentCriteria {
...
 	fn check_assignment_cert(
 		&self,
-		claimed_core_index: Vec<CoreIndex>,
+		claimed_core_index: Option<CoreIndex>,
 		validator_index: ValidatorIndex,
 		config: &Config,
 		relay_vrf_story: RelayVRFStory,
 		assignment: &AssignmentCertV2,
-		// Backing groups for each assigned core `CoreIndex`.
+		// Backing groups for each "leaving core".
```
[the remark this answers is not in the document]
Comment (ordian) [reply node/core/approval-voting/src/criteria.rs:263]: Found some docs: 

> /// The `leaving_cores` parameter indicates all cores within the block where a candidate was included,
> /// as well as the group index backing those.

File: node/core/approval-voting/src/criteria.rs (hunk 9 above)
Comment (ordian) [reply node/core/approval-voting/src/criteria.rs:693]: nvm, we were checking this before

File: node/core/approval-voting/src/approval_db/v1/mod.rs [hunk 11]
```diff
@@ -161,11 +161,16 @@ pub struct Config {
 /// Details pertaining to our assignment on a block.
 #[derive(Encode, Decode, Debug, Clone, PartialEq)]
 pub struct OurAssignment {
+	/// Our assignment certificate.
 	pub cert: AssignmentCertV2,
+	/// The tranche for which the assignment refers to.
 	pub tranche: DelayTranche,
+	/// Our validator index for the session in which the candidate was included.
```
[the remark this answers is not in the document]
Comment (sandreim) [reply node/core/approval-voting/src/approval_db/v1/mod.rs:168]: Refers to all candidates in `assignment_bitfield`.

File: node/core/approval-voting/src/lib.rs (hunk 8 above)
Comment (sandreim) [reply node/core/approval-voting/src/lib.rs:1072]: This fn doesn't really check that, but it is checked when the assignment is imported.

File: node/core/approval-voting/src/criteria.rs (hunk 9 above)
Comment (sandreim) [reply node/core/approval-voting/src/criteria.rs:693]: Yeah, there is also a secondary check for claimed candidate indices because this only checks the crypto part, and can yield cores that did not became free in that block(a candidate was included).

Comment (sandreim) [reply node/core/approval-voting/src/criteria.rs:693]: I'll add a comment

File: node/core/approval-voting/src/lib.rs (hunk 8 above)
Comment (sandreim) [reply node/core/approval-voting/src/lib.rs:1072]: Actually this code is only for our own assignments. `candidates_indices` must be greater or equal to number of 1 bits in the produced `AssignmentBitfield`.

`compute_relay_vrf_modulo_assignments_v2` needs the `leaving_cores` parameter to  filters out invalid core assignments. These are also checked when importing the assignment.

File: node/core/approval-voting/src/approval_db/v1/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ordian) [reply node/core/approval-voting/src/approval_db/v1/mod.rs:168]: I was referring to "the candidate was" part :) Should be "the candidate**s** **were**"?

File: node/primitives/src/approval.rs [hunk 12]
```diff
@@ -31,24 +31,136 @@ use sp_consensus_babe as babe_primitives;
...
+	/// A static context associated with producing randomness for a core.
+	pub const CORE_RANDOMNESS_CONTEXT: &[u8] = b"A&V CORE v2";
+	/// A static context associated with producing randomness for v2 multi-core assignments.
+	pub const ASSIGNED_CORE_CONTEXT: &[u8] = b"A&V ASSIGNED v2";
+	/// A static context used for all relay-vrf-modulo VRFs for v2 multi-core assignments.
+	pub const RELAY_VRF_MODULO_CONTEXT: &[u8] = b"A&V MOD v2";
+	/// A read-only bitvec wrapp to efficiently store candidate or core assignments. Each 1 bit identifies
+	/// a candidate(or core) by the bitfield bit index.
+	#[derive(Clone, Debug, Encode, Decode, Hash, PartialEq, Eq)]
+	pub struct AssignmentBitfield(BitVec<u8, bitvec::order::Lsb0>);
```
Comment (ordian) [inline node/primitives/src/approval.rs:73]: I'd name it either `CandidateBitfield` or `CoreBitfield` to indicate that the bit are addressed by core index.
Typo: wrapper?

File: node/core/approval-voting/src/lib.rs [hunk 13]
```diff
@@ -1785,17 +1830,25 @@ fn check_and_import_assignment(
 		assigned_candidate_hashes.push(assigned_candidate_hash);
 	}
 
+	let claimed_core_index = match assignment.cert.kind {
+		// TODO, near future: remove CoreIndex from certificates completely.
```
Comment (ordian) [inline node/core/approval-voting/src/lib.rs:1834]: Could you elaborate? Will this be addressed in the parent PR?

**Review by ordian — APPROVED** [review]
Looks reasonable to be merged into parent PR. Will take another look at the latter once it's ready.

File: node/core/approval-voting/src/lib.rs (hunk 13 above)
Comment (sandreim) [reply node/core/approval-voting/src/lib.rs:1834]: Yeah, that should be a future PR, not parent. I'll log a separate issue for it.

File: node/primitives/src/approval.rs (hunk 12 above)
Comment (sandreim) [reply node/primitives/src/approval.rs:73]: typo indeed.

It would look strange to use CandidateBitfield for cores. To me it looks good enough as an assignment refers to both of them.

File: node/core/approval-voting/src/approval_db/v1/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (sandreim) [reply node/core/approval-voting/src/approval_db/v1/mod.rs:168]: yeah, I totally missed that :)

File: node/primitives/src/approval.rs (hunk 12 above)
Comment (sandreim) [reply node/primitives/src/approval.rs:73]: I'll try to improve a this a bit
